WebDriverAgentLib/Categories/XCUIElement+FBFind.m

```diff
@@ -83,18 +85,43 @@ - (void)descendantsWithProperty:(NSString *)property value:(NSString *)value par
   // XPath will try to match elements only class name, so requesting elements by XCUIElementTypeAny will not work. We should use '*' instead.
   xpathQuery = [xpathQuery stringByReplacingOccurrencesOfString:@"XCUIElementTypeAny" withString:@"*"];
   NSArray *matchingSnapshots = [self.lastSnapshot fb_descendantsMatchingXPathQuery:xpathQuery];
-  NSArray *allElements = [[self descendantsMatchingType:XCUIElementTypeAny] allElementsBoundByIndex];
-  NSArray *matchingElements = [self filterElements:allElements matchingSnapshots:matchingSnapshots];
+  if (0 == [matchingSnapshots count]) {
+    return @[];
+  }
+  // Prefiltering elements speeds up search by XPath a lot, because [element resolve] is the most expensive operation here
+  NSSet *matchedTypes = [XCUIElement getSnapshotsTypes:matchingSnapshots];
+  NSDictionary *elementsByType = [self splitElementsByType:matchingSnapshots byTypes:matchedTypes];
+  NSArray *matchingElements = [self filterElements:elementsByType matchingSnapshots:matchingSnapshots];
   return matchingElements;
 }
 
-- (NSArray<XCUIElement *> *)filterElements:(NSArray<XCUIElement *> *)elements matchingSnapshots:(NSArray<XCElementSnapshot *> *)snapshots
++(NSSet *)getSnapshotsTypes:(NSArray<XCElementSnapshot *> *)matchingSnapshots
+{
+  NSMutableSet *matchingTypes = [NSMutableSet set];
+  [matchingSnapshots enumerateObjectsUsingBlock:^(XCElementSnapshot *snapshot, NSUInteger snapshotIdx, BOOL *stopSnapshotEnum) {
+    [matchingTypes addObject:@(snapshot.elementType)];
+  }];
+  return matchingTypes;
+}
+
+- (NSDictionary<NSNumber *, NSArray<XCUIElement *> *> *)splitElementsByType:(NSArray<XCElementSnapshot *> *)snapshots byTypes:(NSSet *)types
+{
+  NSMutableDictionary *result = [NSMutableDictionary dictionary];
+  [types enumerateObjectsUsingBlock:^(NSNumber *elementTypeAsNumber, BOOL *stopEnum) {
+    XCUIElementType elementType = (XCUIElementType)elementTypeAsNumber.unsignedIntegerValue;
+    NSArray *descendantsOfType = [[self descendantsMatchingType:elementType] allElementsBoundByIndex];
+    result[elementTypeAsNumber] = descendantsOfType;
+  }];
+  return result.copy;
+}
+
+- (NSArray<XCUIElement *> *)filterElements:(NSDictionary<NSNumber *, NSArray<XCUIElement *> *> *)elementsMap matchingSnapshots:(NSArray<XCElementSnapshot *> *)snapshots
```

Comment:
Make it class method. If possible, write some tests, but probably it will be painful